dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/ExclusionFiles.java

```diff
@@ -0,0 +1,254 @@
+ *   <li>A MethodMatcher matches method symbol specified by className and name attribute.
+ *   <li>A FieldMatcher matches field symbol specified by className and name attribute.
+ * </ul>
+ */
+class ExclusionFiles {
+  private static final XMLEventFactory eventFactory = XMLEventFactory.newInstance();
+
+  private static final QName LINKAGE_CHECKER_FILTER_TAG = QName.valueOf("LinkageCheckerFilter");
+  private static final QName CLASS_TAG = QName.valueOf("Class");
+  private static final QName LINKAGE_ERROR_TAG = QName.valueOf("LinkageError");
+  private static final QName TARGET_TAG = QName.valueOf("Target");
+  private static final QName SOURCE_TAG = QName.valueOf("Source");
+  private static final QName METHOD_TAG = QName.valueOf("Method");
+  private static final QName FIELD_TAG = QName.valueOf("Field");
+
+  static ImmutableList<LinkageErrorMatcher> parse(Path exclusionFile)
+      throws SAXException, IOException, VerifierConfigurationException {
+
+    InputSource inputSource = new InputSource(Files.newInputStream(exclusionFile));
+    inputSource.setSystemId(exclusionFile.toUri().toString());
+
+    return parse(inputSource);
+  }
+
+  static ImmutableList<LinkageErrorMatcher> parse(URL exclusionFile)
+      throws SAXException, IOException, VerifierConfigurationException {
+
+    InputSource inputSource = new InputSource(exclusionFile.openStream());
+    inputSource.setSystemId(exclusionFile.toString());
+
+    return parse(inputSource);
+  }
+
+  private static ImmutableList<LinkageErrorMatcher> parse(InputSource inputSource)
+      throws SAXException, IOException, VerifierConfigurationException {
+
+    XMLReader reader = createXmlReader();
+
+    ExclusionFileHandler handler = new ExclusionFileHandler();
+    reader.setContentHandler(handler);
+
+    reader.parse(inputSource);
+
+    return handler.getMatchers();
+  }
+
+  private static XMLReader createXmlReader()
+      throws SAXException, IOException, VerifierConfigurationException {
+    // Validate and parse XML files in one pass using Jing validator as a filter.
+    // http://iso-relax.sourceforge.net/JARV/JARV.html#use_42
+    VerifierFactory factory = VerifierFactory.newInstance("http://relaxng.org/ns/structure/1.0");
+    InputStream linkageCheckerSchema =
+        ExclusionFiles.class.getClassLoader().getResourceAsStream("linkage-checker-exclusion.rng");
+    Schema schema = factory.compileSchema(linkageCheckerSchema);
+    Verifier verifier = schema.newVerifier();
+
+    // DraconianErrorHandler throws SAXException upon invalid structure
+    verifier.setErrorHandler(new DraconianErrorHandler());
+    VerifierFilter filter = verifier.getVerifierFilter();
+    filter.setParent(XMLReaderFactory.createXMLReader());
+    return filter;
+  }
+
+  /** Writes {@code linkageErrors} as exclusion rules into {@code outputFile}. */
+  static void write(Path outputFile, Multimap<SymbolProblem, ClassFile> linkageErrors)
```

Comment:
This is the sort of use of multimap that confuses me. Is the classFile the source or the target? It's unclear how it's encoded or what it means. This should be a Collection<SymbolProblem> and the symbol problem should know the source and the target. What we're writing into the file is a list of linkage errors. Why aren't we passing a list of linkage errors into this method?